Mathlib/Tactic/MoveAdd.lean

```diff
@@ -0,0 +1,427 @@
+/-  output:
+`([0, 0], [0])`
+the output LHS list `[0, 0]` consists of the first `0` and the `MVarId`.
+the output RHS list `[0]` corresponds to the last `0` in `L`.
+-/
+```
+-/
+def pairUp : List (Expr × Bool × Syntax) → List Expr →
+    MetaM ((List (Expr × Bool)) × List (Expr × Bool × Syntax))
+  | (m::ms), l => do
+    match ← l.findM? (isDefEq · m.1) with
+      | none => let (found, unfound) := ← pairUp ms l; return (found, m::unfound)
+      | some d => let (found, unfound) := ← pairUp ms (l.erase d)
+                  return ((d, m.2.1)::found, unfound)
+  | _, _ => return ([], [])
+
+open Elab.Tactic
+
+/-- `move_oper_simpCtx` is the `Simp.Context` for the reordering internal to `move_oper`.
+To support a new binary operation, extend the list in this definition, so that it contains
+enough lemmas to allow `simp` to close a generic permutation goal for the new binary operation.
+-/
+def move_oper_simpCtx : MetaM Simp.Context := do
+  -- the `simpOnlyBuiltins` are `eq_self` and `iff_self`
+  let simpNames := simpOnlyBuiltins ++ [
+    ``add_comm, ``add_assoc, ``add_left_comm,  -- for `HAdd.hAdd`
+    ``mul_comm, ``mul_assoc, ``mul_left_comm,  -- for `HMul.hMul`
+    ``and_comm, ``and_assoc, ``and_left_comm,  -- for `and`
+    ``or_comm,  ``or_assoc,  ``or_left_comm,   -- for `or`
+    ``max_comm, ``max_assoc, ``max_left_comm,  -- for `max`
+    ``min_comm, ``min_assoc, ``min_left_comm   -- for `min`
+    ]
+  let simpThms := ← simpNames.foldlM (·.addConst ·) ({} : SimpTheorems)
+  return { simpTheorems := #[simpThms] }
+
+/-- `reorderAndSimp mv op instr` takes as input an `MVarId`  `mv`, the name `op` of a binary
+operation and a list of "instructions" `instr` that it passes to `permuteExpr`.
+
+* It creates a version `permuted_mv` of `mv` with subexpressions representing `op`-sums reordered
+  following `instructions`.
+* It produces 2 temporary goals by applying `Eq.mpr` and unifying the resulting meta-variable with
+  `permuted_mv`: `[⊢ mv = permuted_mv, ⊢ permuted_mv]`.
+* It tries to solve the goal `mv = permuted_mv` by a simple-minded `simp` call, using the
+  `op`-analogues of `add_comm, add_assoc, add_left_comm`.
+-/
+def reorderAndSimp (mv : MVarId) (op : Name) (instr : List (Expr × Bool)) :
+    MetaM (List MVarId) := do
+  let permExpr := ← permuteExpr op (← mv.getType'') instr
+  -- generate the implication `permutedMv → mv = permutedMv → mv`
+  let eqmpr := ← mkAppM ``Eq.mpr #[← mkFreshExprMVar (← mkEq (← mv.getType) permExpr)]
+  let twoGoals := ← mv.apply eqmpr
+  guard (twoGoals.length == 2) <|>
+    throwError m!"There should only be 2 goals, instead of {twoGoals.length}"
+  -- `permGoal` is the single goal `mv_permuted`, possibly more operations will be permuted later on
+  let permGoal := ← twoGoals.filterM fun v => return !(← v.isAssigned)
+  match ← (simpGoal (permGoal[1]!) (← move_oper_simpCtx)) with
+    | (some x, _) => throwError m!"'move_oper' could not solve {indentD x.2}"
+    | (none, _) => return permGoal
```

Comment:
```suggestion
  let permExpr ← permuteExpr op (← mv.getType'') instr
  -- generate the implication `permutedMv → mv = permutedMv → mv`
  let eqmpr ← mkAppM ``Eq.mpr #[← mkFreshExprMVar (← mkEq (← mv.getType) permExpr)]
  let twoGoals ← mv.apply eqmpr
  guard (twoGoals.length == 2) <|>
    throwError m!"There should only be 2 goals, instead of {twoGoals.length}"
  -- `permGoal` is the single goal `mv_permuted`, possibly more operations will be permuted later on
  let permGoal ← twoGoals.filterM fun v => return !(← v.isAssigned)
  match ← (simpGoal (permGoal[1]!) (← move_oper_simpCtx)) with
    | (some x, _) => throwError m!"'move_oper' could not solve {indentD x.2}"
    | (none, _) => return permGoal
```